Framework/PythonInterface/mantid/utils/absorptioncorrutils.py

```diff
@@ -32,174 +33,215 @@ def _getBasename(filename):
-        # disk -> memory
+    # step_1: check memory
+    if mtd.doesExist(abs_wksp_sample):
+        found_abs_wksp_sample = mtd[abs_wksp_sample].run()["absSHA1"].value == sha1
+    if mtd.doesExist(abs_wksp_container):
+        found_abs_wksp_container = mtd[abs_wksp_container].run()["absSHA1"].value == sha1
+
+    # step_2: load from disk if either is not found in memory
+    if (not found_abs_wksp_sample) or (not found_abs_wksp_container):
         if os.path.exists(cache_file_name):
-            # LoadNexus(Filename=cache_file_name, OutputWorkspace=wsn_as)
-            # if mtd[wsn_as].id() == "WorkspaceGroup":
-            #     UnGroupWorkspace(InputWorkspace=wsn_as)
-            wsntmp = f"{wsn_as}_wsg"
+            wsntmp = f"tmpwsg"
             Load(Filename=cache_file_name, OutputWorkspace=wsntmp)
             wstype = mtd[wsntmp].id()
             if wstype == "Workspace2D":
-                RenameWorkspace(InputWorkspace=wsntmp, OutputWorkspace=wsn_as)
+                RenameWorkspace(InputWorkspace=wsntmp, OutputWorkspace=abs_wksp_sample)
             elif wstype == "WorkspaceGroup":
-                # there should be exactly two workspaces inside, ungroup them will
-                # restore them with the orignal name (wsn_as, wsn_ac)
                 UnGroupWorkspace(InputWorkspace=wsntmp)
             else:
                 raise ValueError(f"Unsupported cached workspace type: {wstype}")
 
-        # now check the memory again
-        # -- wsn_as exist for all three methods
-        # NOTE: The new attribute absSHA1 should be there.
-        if mtd.doesExist(wsn_as):
-            found_as = mtd[wsn_as].run()["absSHA1"].value == sha1
-        if mtd.doesExist(wsn_ac):
-            found_ac = mtd[wsn_ac].run()["absSHA1"].value == sha1
+    # step_3: check memory again
+    if mtd.doesExist(abs_wksp_sample):
+        found_abs_wksp_sample = mtd[abs_wksp_sample].run()["absSHA1"].value == sha1
+    if mtd.doesExist(abs_wksp_container):
+        found_abs_wksp_container = mtd[abs_wksp_container].run()["absSHA1"].value == sha1
+
+    return found_abs_wksp_sample, found_abs_wksp_container, abs_wksp_sample, abs_wksp_container
+
+
+# NOTE:
+#  In order to use the decorator, we must have consistent naming
+#  or kwargs as this is probably the most reliable way to get
+#  the desired data piped in multiple location
+#  -- bare minimum signaure of the function
+#    func(wksp_name: str, abs_method:str, cache_dir="")
+def abs_cache(func):
+    """decorator to make the caching process easier
+    
+    example:
+    without caching:
+        SNSPowderReduction successful, Duration 5 minutes 53.54 seconds
+    with caching (disk):
+        SNSPowderReduction successful, Duration 1 minutes 14.18 seconds
+    Speedup by
+        4.7660x
+    """
+    @wraps(func)
+    def inner(*args, **kwargs):
+        """
+        How caching name works
+        """
+        # unpack key arguments
+        wksp_name = args[0]
```

Comment:
It is possible, but I don't believe the decorator function is the best place to validate if workspace exists.
When the function it decorates is called  (in this case, `calc_absorption_corr_using_wksp`), the caller of the decorated function should be responsible of making sure a valid workspace name is used.

Reply:
You're assuming `args[0]` is the name of an existing workspace because it's being used as such in the body of the inner function itself, irrespective of how it is used in the function to be decorated. If you don't want to insert a validator then at the very least there should be a description in the docstring of the decorator, making it plain that the first  item in the positional arguments of the function to be decorated is the name of an existing workspace.

Reply:
- `args` is the list of the positional arguments, and `kwargs` is the list of keyword arguments.
- the decorated function,  `calc_absorption_corr_using_wksp`, sets the first positional argument to be the name of the workspace, and the decorator is passively acknowledging this.

Also, I think there is some misunderstanding here.
The decorated function is not using the decorator (in fact, it doesn't even know if there is a decorator or not).
On the contrary, it is the decorator function using the function it is decorating.
In this case, the decorator takes in `calc_absorption_corr_using_wksp`, then add cache checking before `calc_absorption_corr_using_wksp` and cache saving after `calc_absorption_corr_using_wksp`.

Reply:
The decorator is making certain assumptions about the functions it will decorate. In particular, it assumes the first two arguments in the signature of the decorated function are of a certain type. This should be either documented or validated.

Reply:
I see your point.
I will update the comments accordingly.